part-1.py

```diff
@@ -4,17 +4,48 @@
 # the appropriate comment.
 
 # factorial
-
+def factorial(n):
+        if n < 0:
+            raise ValueError("ValueError exception thrown")
+        if n == 0 :
+            return 1
+        else:
+            return n * factorial(n -1 )
+    
+    
 
 
 # reverse
-
+def reverse(text):
+    if text:
+        text = text[::-1]
+        return text
+    if text == "":
+        return ""
+    
 
 
 # bunny
-
+def bunny(count):
+    if count < 1:
+        return 0
+    return 2 + bunny(count - 1)
 
 
 # is_nested_parens
-
+def is_nested_parens(parens, n = 0):
+    if parens == "":
+        return True
+    if len(parens) % 2 == 1:
+        return False
+    
+    parens_reversed = parens[::-1]
+    
+    if n >= len(parens):
+        return True
+        
+    if parens[n] == parens_reversed[n]:
+        return False
+    
+    return is_nested_parens(parens, n +1)
```

Comment:
so you are calling this recursively, but `parens[::-1]` is doing all the work. It is calling a for loop and re-concatenating the string.

What we want to do is have our function check the starting index and the last index to see if they are equal. If they are, we call `is_nested_parens` again and decrement the last index, increment the starting index and then check again.